astropy/io/fits/fitsrec.py

```diff
@@ -1166,6 +1166,12 @@ def _scale_back(self, update_heap_pointers=True):
                            np.array([ord('T')], dtype=np.int8)[0])
                 raw_field[:] = np.choose(field, choices)
 
+        if isinstance(recformat, _FormatP):
+            P_heap_limit = ((2**32) / 2) - 1
+            if recformat._format_code == 'P' and heapsize > P_heap_limit:
```

Comment:
Would it be possible to switch automatically to Q format on writing?

Also detail: I'd just do `if recformat._format_code='P' and heapsize >= 2**31`

Reply:
> Would it be possible to switch automatically to Q format on writing?

I will check... I need some more time to understand the code...

Reply:
Not the easiest code in astropy, I fear...

Note that perhaps the best solution is to first go with the fix you have here, which will at least alert the user to the problem; nobody will complain if we later make it automatic!

Reply:
I might be wrong but my understanding is that the heapsize is a running quantity: I cannot figure out how I could come back and recast part of the table once it is already set. Sorry, this is above my python skills (at least for now... :) ).

Reply:
OK, that makes sense. Probably best to keep it simple for now; with your change, at least the user has a chance to know what is going on!